src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,162 @@
+            );
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (!this.options.initalState) {
+            console.warn('The stateful action button cannot render if the inital states is not defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (this.options.states.indexOf(this.options.initalState) < 0) {
+            console.warn('The stateful action button cannot render if the initial state is not in the list of states.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        return !this.options.allowedTransitions || this.areTransitionsValid();
+    }
+
+    private areTransitionsValid(): boolean {
+        for (let index = 0; index < this.options.allowedTransitions.length; index++) {
+            const transition = this.options.allowedTransitions[index];
+            if (this.options.states.indexOf(transition.from) == -1) {
+                console.warn(this.generateInvalidTransitionMessage(index, true));
+                return false;
+            }
+            if (this.options.states.indexOf(transition.from) == -1) {
+                console.warn(this.generateInvalidTransitionMessage(index, false));
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private generateInvalidTransitionMessage(transitionNumber: number, isOrigin: boolean) {
+        return `The stateful action button cannot render if its origin is not in the list of states:\n\t${
+            isOrigin ? 'Origin' : 'Destination'
+        } of Transition #${transitionNumber} is not in the list of states. Ensure to use the same object reference as in the options.states.`;
+    }
+
+    /**
+     * Check if a transition from the current state to @param state is allowed.
+     * @param state the destination of the transition
+     */
+    private isTransitionAllowed(state: IStatefulActionButtonState) {
+        return (
+            this.options.states.indexOf(state) > -1 &&
+            (!this.options.allowedTransitions ||
```

Comment:
You can use the look ahead operator here.
`his.options.allowedTransitions?.some`